reference/latest/src/client/java/com/example/docs/datagen/FabricDocsReferenceModelProvider.java

```diff
@@ -0,0 +1,161 @@
+import net.minecraft.client.data.ModelIds;
+import net.minecraft.client.data.TextureKey;
+import net.minecraft.client.data.TextureMap;
+import net.minecraft.client.data.TexturedModel;
+import net.minecraft.client.data.VariantSetting;
+import net.minecraft.client.data.VariantSettings;
+import net.minecraft.client.data.VariantsBlockStateSupplier;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.math.Direction;
+
+import java.util.Optional;
+
+public class FabricDocsReferenceModelProvider extends FabricModelProvider {
+	public FabricDocsReferenceModelProvider(FabricDataOutput output) {
+		super(output);
+	}
+
+	@Override
+	public void generateBlockStateModels(BlockStateModelGenerator blockStateModelGenerator) {
+		// :::datagen-model:provider
+
+		// :::datagen-model:cube-all
+		blockStateModelGenerator.registerSimpleCubeAll(ModBlocks.STEEL_BLOCK);
+		// :::datagen-model:cube-all
+
+		// :::datagen-model:cube-top-for-ends
+		blockStateModelGenerator.registerSingleton(ModBlocks.PIPE_BLOCK, TexturedModel.END_FOR_TOP_CUBE_COLUMN);
+		// :::datagen-model:cube-top-for-ends
+
+		// :::datagen-model:block-texture-pool-normal
+		blockStateModelGenerator.registerCubeAllModelTexturePool(ModBlocks.RUBY_BLOCK)
+				.stairs(ModBlocks.RUBY_STAIRS)
+				.slab(ModBlocks.RUBY_SLAB)
+				.fence(ModBlocks.RUBY_FENCE);
+		// :::datagen-model:block-texture-pool-normal
+
+		// :::datagen-model:door-and-trapdoor
+		blockStateModelGenerator.registerDoor(ModBlocks.RUBY_DOOR);
+		blockStateModelGenerator.registerTrapdoor(ModBlocks.RUBY_TRAPDOOR);
+		// blockStateModelGenerator.registerOrientableTrapdoor(ModBlocks.RUBY_TRAPDOOR);
+		// :::datagen-model:door-and-trapdoor
+
+		// :::datagen-model-custom:method-call
+		CustomBlockStateModelGenerator.registerVerticalSlab(
+				blockStateModelGenerator,
+				ModBlocks.VERTICAL_OAK_LOG_SLAB,
+				Blocks.OAK_LOG,
+				CustomBlockStateModelGenerator.blockAndTopForEnds(Blocks.OAK_LOG)
+		);
+		// :::datagen-model-custom:method-call
+
+
+
+		// :::datagen-model:provider
+	}
+	// :::datagen-model:provider
+
+
+
+	// used just for examples, not for actual data generation
+	@SuppressWarnings("unused")
+	public void exampleBlockStateGeneration(BlockStateModelGenerator blockStateModelGenerator) {
+		// :::datagen-model:block-texture-pool-family
+		blockStateModelGenerator.registerCubeAllModelTexturePool(ModBlocks.RUBY_BLOCK).family(ModBlocks.RUBY_FAMILY);
+		// :::datagen-model:block-texture-pool-family
+	}
+
+	// :::datagen-model:provider
+
+	@Override
+	public void generateItemModels(ItemModelGenerator itemModelGenerator) {
+		// :::datagen-model:provider
+
+		//TODO Since I have little experience with generating item models, I will leave this to someone more experienced (Fellteros)
+
+		// :::datagen-model:provider
+	}
+
+	// :::datagen-model:provider
+
+
+	// Inner class containing all Objects needed for the custom datagen tutorial.
+	public static class CustomBlockStateModelGenerator {
```

Comment:
perhaps it should be mentioned in the article that we have an inner helper class?